src/sentry/static/sentry/app/utils/projects.jsx

```diff
@@ -0,0 +1,296 @@
+
+    this.setState({
+      fetching: true,
+    });
+
+    try {
+      const {results, hasMore} = await fetchProjects(api, orgId, {limit});
+
+      this.setState({
+        fetching: false,
+        fetchedProjects: results,
+        initiallyLoaded: true,
+        hasMore,
+      });
+    } catch (err) {
+      console.error(err); // eslint-disable-line no-console
+
+      this.setState({
+        fetching: false,
+        fetchedProjects: [],
+        initiallyLoaded: true,
+        fetchError: err,
+      });
+    }
+  };
+
+  /**
+   * This is an action provided to consumers for them to update the current projects
+   * result set using a simple search query. You can allow the new results to either
+   * be appended or replace the existing results.
+   *
+   * @param {String} search The search term to use
+   * @param {Object} options Options object
+   * @param {Boolean} options.append Results should be appended to existing list (otherwise, will replace)
+   */
+  handleSearch = async (search, {append} = {}) => {
+    const {api, orgId, limit} = this.props;
+
+    this.setState({fetching: true});
+
+    try {
+      const {results, hasMore} = await fetchProjects(api, orgId, {search, limit});
+
+      this.setState(state => {
+        let fetchedProjects;
+        if (append) {
+          // Remove duplicates
+          fetchedProjects = uniqBy(
+            [...state.fetchedProjects, ...results],
+            ({slug}) => slug
+          );
+        } else {
+          fetchedProjects = results;
+        }
+        return {
+          fetchedProjects,
+          hasMore,
+          fetching: false,
+        };
+      });
+    } catch (err) {
+      console.error(err); // eslint-disable-line no-console
+
+      this.setState({
+        fetching: false,
+        fetchError: err,
+      });
+    }
+  };
+
+  render() {
+    const {slugs, children} = this.props;
+
+    return children({
+      // May not need to expose this?
+      initiallyLoaded: this.state.initiallyLoaded,
+
+      // We want to make sure that at the minimum, we return a list of objects with only `slug`
+      // while we load actual project data
+      projects: this.state.initiallyLoaded
+        ? [...this.state.fetchedProjects, ...this.state.projectsFromStore]
+        : (slugs && slugs.map(slug => ({slug}))) || [],
+
+      // This is set when we fail to find some slugs from both store and API
+      isIncomplete: this.state.isIncomplete,
+
+      // This is state for when fetching data from API
+      fetching: this.state.fetching,
+
+      // Project results (from API) are paginated and there are more projects
+      // that are not in the initial queryset
+      hasMore: this.state.hasMore,
+
+      // Calls API and searches for project, accepts a callback function with signature:
+      //
+      // fn(searchTerm, {append: bool})
+      onSearch: this.handleSearch,
+    });
+  }
+}
+
+export default withProjects(withApi(Projects));
```

Comment:
If this component has `withProjects()` on it how likely will it be that we won't have all projects in the client?

Reply:
As it is now... very unlikely it does not have all projects.